p2p/src/secret_connection.rs

```diff
@@ -42,38 +42,59 @@ pub const DATA_MAX_SIZE: usize = 1024;
 const DATA_LEN_SIZE: usize = 4;
 const TOTAL_FRAME_SIZE: usize = DATA_MAX_SIZE + DATA_LEN_SIZE;
 
-/// Encrypted connection between peers in a Tendermint network
-pub struct SecretConnection<IoHandler: Read + Write + Send + Sync> {
-    io_handler: IoHandler,
-    protocol_version: Version,
-    recv_nonce: Nonce,
-    send_nonce: Nonce,
+/// Handshake states
+
+/// AwaitingEphKey means we're waiting for the remote ephemeral pubkey.
+struct AwaitingEphKey {
+    local_privkey: ed25519::Keypair,
+    local_eph_privkey: Option<EphemeralSecret>,
+}
+
+/// AwaitingAuthSig means we're waiting for the remote authenticated signature.
+struct AwaitingAuthSig {
+    sc_mac: [u8; 32],
+    kdf: Kdf,
     recv_cipher: ChaCha20Poly1305,
     send_cipher: ChaCha20Poly1305,
-    remote_pubkey: Option<PublicKey>,
-    recv_buffer: Vec<u8>,
+    local_signature: ed25519::Signature,
 }
 
-impl<IoHandler: Read + Write + Send + Sync> SecretConnection<IoHandler> {
-    /// Returns authenticated remote pubkey
-    pub fn remote_pubkey(&self) -> PublicKey {
-        self.remote_pubkey.expect("remote_pubkey uninitialized")
-    }
+/// Handshake is a process of establishing the SecretConnection between two peers.
+/// Specification: https://github.com/tendermint/spec/blob/master/spec/p2p/peer.md#authenticated-encryption-handshake
+struct Handshake<S> {
+    protocol_version: Version,
+    state: S,
```

Comment:
One significant change we should land is to have `S` be constrained. Otherwise at the moment a `Handshake` can be constructed with any `S`. Commonly the `State` trait is sealed away in a private module, together with the impls to make it impossible for it to be constructed from the outside with anything but a sanctioned list of states. For reference:

https://github.com/informalsystems/tendermint-rs/blob/6cfeeff8ff58d2cf58f5dc40a60dc356db9504c2/p2p/src/peer.rs#L171-L182

Reply:
> from the outside with anything but a sanctioned list of states

but Handshake is not exposed. How one can construct it from the outside?

Reply:
The example from above would only be meaningful, if Handshake would be allowed to be publicly constructed, so we can disregard.